tests/jdom/tester.ts

```diff
@@ -0,0 +1,219 @@
+import { JDBus } from "../../src/jdom/bus"
+import JDServiceProvider from "../../src/jdom/serviceprovider"
+import { EVENT, DEVICE_ANNOUNCE } from "../../src/jdom/constants"
+import { mkBus } from "../testutils"
+
+import { JDEvent } from "../../src/jdom/event"
+import { JDDevice } from "../../src/jdom/device"
+import JDServiceServer from "../../src/jdom/serviceserver"
+import { assert } from "../../src/jdom/utils"
+import { JDService } from "../../src/jdom/service"
+
+// Set equals is not a built-in operation.
+function setEquals<T>(set1: Set<T>, set2: Set<T>): boolean {
+    if (set1.size != set2.size) {
+        return false
+    }
+    set1.forEach(value => {
+        if (!set2.has(value)) {
+            return false
+        }
+    })
+    return true
+}
+
+// Creates a test bus, runs the test body function, and tears down the test bus.
+// Bus setup should be handled within the test body
+export async function withBus(test: (bus: JDBus) => Promise<void>) {
+    const bus = mkBus()
+
+    bus.start()
+
+    // Actually run the test here
+    await test(bus)
+
+    bus.stop()
+}
+
+// Structure returned from createServices for each server, that contains the server passed in,
+// the created device, and the service on the device corresponding to the server.
+export interface CreatedServerService<ServiceType extends JDServiceServer> {
+    server: ServiceType
+    device: JDDevice
```

Comment:
you can always recover the device from service (``service.device``) so this one might be not needed

Reply:
Sure, removed for now since it's not used.